[Refactoring] Remove the backend global singleton

NicolasDorier opened the pull request:
This is needed for https://github.com/zkSNACKs/WalletWasabi/pull/1815

I remove the backend global singleton by initializing it as part of the `Startup` and registering it in the service collections of the website so it can be injected where needed.

This has the nice side effect to effectively test the initialization code that was previously in `Program.cs`.
I needed to add `ListenRoundConfigFileChanges` and disable it for the tests to pass.

Also I removed some mess to detect the `wwwroot` directory and indeed just use `HostingEnvironment.WebRootPath` provided by asp.net core.

nopara73 commented:
Idea ACK. Before code review, let me ask, is it the same pattern you used in the Gui side?

NicolasDorier commented:
Yes, but actually it is simpler than the Gui side, because Avalonia IoC system is a big hack, while ASP.NET Core is made for IoC.

nopara73 commented on WalletWasabi.Tests/XunitConfiguration/RegTestFixture.cs:
```diff
@@ -31,26 +32,36 @@ public RegTestFixture()
 			var rpc = BackendRegTestNode.CreateRpcClient();
-
-			var config = new Config(rpc.Network, rpc.Authentication, IPAddress.Loopback.ToString(), IPAddress.Loopback.ToString(), BackendRegTestNode.Endpoint.Address.ToString(), Network.Main.DefaultPort, Network.TestNet.DefaultPort, BackendRegTestNode.Endpoint.Port);
+			var connectionString = new NBitcoin.RPC.RPCCredentialString() 
```
Can you using NBitcoin.RPC instead of writing out the full name?

nopara73 commented on WalletWasabi.Tests/XunitConfiguration/RegTestFixture.cs:
```diff
@@ -31,26 +32,36 @@ public RegTestFixture()
 			Task.WaitAny(delayTask, hostInitializationTask); // Wait for server to initialize (Without this OSX CI will fail)
 		}
-
+		public Backend.Global Global { get; set; }
```
Could you put it in front?

NicolasDorier replied:
you mean move the property up?

nopara73 replied:
Yes. Done.

nopara73 commented on WalletWasabi.Backend/UnversionedWebBuilder.cs:
```diff
@@ -41,7 +44,7 @@ public static void CloneAndUpdateOnionIndexHtml()
 		}
 
-		public static void UpdateCoinJoinsHtml(IEnumerable<string> coinJoins)
+		public void UpdateCoinJoinsHtml(Global backendGlobal, IEnumerable<string> coinJoins)
```
Same comment here.

nopara73 commented on WalletWasabi.Backend/Startup.cs:
```diff
@@ -80,34 +81,34 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
 		}
 
-		private void OnShutdown()
+		private void OnShutdown(Global backendGlobal)
```
Same here.

nopara73 commented on WalletWasabi.Backend/Startup.cs:
```diff
@@ -80,34 +81,34 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env)
 		}
 
-		private async Task CleanupAsync()
+		private async Task CleanupAsync(Global backendGlobal)
```
Same here.

NicolasDorier commented:
@nopara73 about `ListenRoundConfigFileChanges`, the reason is very simple.

I moved a bunch of code out of the `Program.cs` into the `StartupTask`. Part of this code is listening the config file changes.

In the regtestsfixture, you had your own boilerplate, similar to what was in the Program.cs but without the listening of config file change.

When I moved all the initialization code out of Program.cs and RegtestFixture, and inside the `StartupTask`, now the Regtests were also listening config file changes.

However, this conflict with the tests. Because the tests change the config at runtime, and the listener would see that the value at runtime is different from the file and reload the file.

An alternative is to get ride of the runtime's `UpdateConfigRound` and directly edit the file. This would have the advantage to also test the config listener.

NicolasDorier commented on WalletWasabi.Backend/Config.cs:
```diff
@@ -43,6 +44,10 @@ public class Config : IConfig
 
+		[DefaultValue(true)]
+		[JsonProperty(PropertyName = "ListenRoundConfigFileChanges", DefaultValueHandling = DefaultValueHandling.Populate)]
+		public bool ListenRoundConfigFileChanges { get; set; }
```
What do you mean? it is the same pattern.

nopara73 replied:
We don't use `[DefaultValue(true)]` here yet.

NicolasDorier commented on WalletWasabi.Backend/Config.cs:
```diff
@@ -208,7 +216,10 @@ public async Task<bool> CheckFileChangeAsync()
 				return true;
 			}
-
+			if (ListenRoundConfigFileChanges != config.ListenRoundConfigFileChanges)
```
I don't understand here, what you suggest would have two lines in a row, which codemaid don't want

nopara73 replied:
I added CodeMaid configuration we are using to the solution file. This will make your CodeMaid extension pick up our rules automatically.

nopara73 replied:
It's more relaxed then the default CodeMaid configuration.

nopara73 replied:
So you can merge the master and fix this.

nopara73 replied:
Oh this code has been obsolated by a commit to the master.

NicolasDorier commented on WalletWasabi.Backend/InitConfigStartupTask.cs:
```diff
@@ -0,0 +1,106 @@
+				Logger.LogWarning(ex, nameof(Program));
+			}
+		}
+	}
```
it is

NicolasDorier commented:
Rebased following the changes you suggested.

nopara73 commented on WalletWasabi.Backend/Extensions.cs:
```diff
@@ -0,0 +1,22 @@
+			=> services.AddTransient<IStartupTask, T>();
+		public static async Task RunWithTasksAsync(this IWebHost webHost, CancellationToken cancellationToken = default)
+		{
+			await Task.WhenAll(webHost.Services.GetServices<IStartupTask>().Select(t => t.ExecuteAsync(cancellationToken)).ToArray());
```
`Task.WhenAll` accepts `IEnumerable`.

```suggestion
			await Task.WhenAll(webHost.Services.GetServices<IStartupTask>().Select(t => t.ExecuteAsync(cancellationToken)));
```

nopara73 commented:
Thanks @NicolasDorier I take this from here.

Your reasoning of `ListenRoundConfigFileChanges` makes sense.

> I find it weird you don't use .NET `FileSystemWatcher`

Never heard of it, but I'll keep it in mind. It sounds very useful. @molnard take a note, too.

> An alternative is to get rid of the runtime's UpdateConfigRound and directly edit the file. This would have the advantage to also test the config listener.

Yes, that's the clean way to do this. I'll implement that.

nopara73 commented:
I removed `ListenRoundConfigFileChanges` and ToFile-ing for the CcjRoundConfig modifications. This is less intrusive change.

However this PR has multiple issues.

# 1 

If I run this PR on the RegTest, then the swagger doesn't work.

![image](https://user-images.githubusercontent.com/9156103/60894143-6e262880-a262-11e9-9c8c-4f5915a63440.png)


# 2

Running the the RegTets puts trash in the Config.cs

```
{
  "Network": "RegTest",
  "BitcoinRpcConnectionString": "server=http://127.0.0.1:11557/;d001e8490cd5428eb7a1a009f6ea442c3097e890:d001e8490cd5428eb7a1a009f6ea442c3097e890",
  "MainNetBitcoinCoreHost": "127.0.0.1",
  "TestNetBitcoinCoreHost": "127.0.0.1",
  "RegTestBitcoinCoreHost": "127.0.0.1",
  "MainNetBitcoinCorePort": 8333,
  "TestNetBitcoinCorePort": 18333,
  "RegTestBitcoinCorePort": 10787
}
```

It messes up the `server` entry and the `RegTestBitcoinCorePort` gets some strange random value.

NicolasDorier commented:
@nopara73 it is not garbage. 

This PR is making sure that each tests are isolated by creating a separate regtest node for each test, both ports are chosen randomly.

The swagger thing is strange, will take a look after rebasing.

NicolasDorier commented:
I rebased and cleaned up the history a bit.

NicolasDorier commented:
@nopara73 I know why it is like that. Because the fixture and tests were communicating via the shared global instance before. I changed that by saving into file, but I should not have done that way. Fixing it.

NicolasDorier commented:
@nopara73 so I found out why I did this actually.
Because now the `Startup` is responsible for creating the `global` instance, you can't just create from outside. So the way I worked around this, is by changing the config file before starting the host.

But because your node builder is a singleton, it is using the `%APPDATA%` one.
So it means that I can't fix this without also removing the test singleton issues.

Another way would be to use `NBitcoin.TestFramework` for the creation of Nodes. It does not suffer from this problem as it does not rely on singletons.

NicolasDorier commented:
wait actually I have another idea to fix it.

NicolasDorier commented:
@nopara73 sorry for that, it was easy to fix thanks to this PR actually: Just have to tell to `Global` to not use the user's config directory and instead create one for the tests. It works nice now.

NicolasDorier commented:
Anything missing now?

nopara73 commented:
Nicolas, I still get the swagger error. Was that supposed to be fixed by now?

- [x] Up to date with the latest master.
- [x] CodeFactor
- [x] CodeMaid
- [x] CI
- [x] Tests pass.
- [x] Tests don't ruin the config.
- [ ] Swagger error fixed.
- [ ] Test behavior.
- [ ] Review code.

NicolasDorier commented:
@nopara73 I think the swagger is because I trahsed your config in %APPDATA%.

Will check now.

NicolasDorier commented:
Just fixed swagger, I don't think it comes from this PR though as you can see.

lontivero commented on WalletWasabi.Backend/Startup.cs:
```diff
@@ -50,14 +58,15 @@ public void ConfigureServices(IServiceCollection services)
 
 			services.AddSingleton<IExchangeRateProvider>(new ExchangeRateProvider());
-
+			services.AddSingleton<Global>(new Global(Configuration["datadir"]));
```
Why do we need getting the `datadir` from a different config file? Is it for testing?

NicolasDorier replied:
yes, so `Regtests` does not use the same config file than the user's datadir.

lontivero commented:
I've reviewed it and everything looks good and pretty clear but it is a big PR. How should we test it?

NicolasDorier commented:
@lontivero run the tests, and run the daemon.

NicolasDorier commented:
Can somebody kick in the box to make wasabi.osx green?

lontivero commented:
Swagger works okay. Tests run okay most of the times (this is not new in my env). The error that I find weird running the tests is this one:

```
2019-07-12 13:24:56 DEBUG IndexBuilderService: NBitcoin.RPC.RPCException: Block height out of range
   at NBitcoin.RPC.RPCResponse.ThrowIfError()
   at NBitcoin.RPC.RPCClient.SendCommandAsyncCore(RPCRequest request, Boolean throwIfRPCError)
   at NBitcoin.RPC.RPCClient.SendCommandAsync(RPCRequest request, Boolean throwIfRPCError)
   at NBitcoin.RPC.RPCClient.GetBlockHashAsync(Int32 height)
   at NBitcoin.RPC.RPCClient.GetBlockAsync(Int32 height)
   at WalletWasabi.Services.IndexBuilderService.<Synchronize>b__35_0() in /home/lontivero/GitHub/WalletWasabi/WalletWasabi/Services/IndexBuilderService.cs:line 261
```

I didn't see this error before.

NicolasDorier commented:
@lontivero I think later I will make each of those tests run a separate bitcoin node. This is probably why the tests are non deterministic.

nopara73 commented:
I will give another round of test and code review to this PR, if it won't be sufficient, I'll apply our new refactoring guidelines here: https://github.com/zkSNACKs/WalletWasabi/blob/master/CONTRIBUTING.md#refactoring

nopara73 commented:
@molnard and me spent 2-3 hours to review this again. We are 99% sure that it won't result in unexpected issues. Merging.
